ospf6d/ospf6_auth_trailer.c

```diff
@@ -0,0 +1,820 @@
+					  oi->interface->name,
+					  ospf6_message_type(oh->type));
+			oi->at_data.rx_drop++;
+			return -1;
+		}
+	} else if (CHECK_FLAG(oi->at_data.flags,
+			      OSPF6_AUTH_TRAILER_MANUAL_KEY)) {
+		auth_str = oi->at_data.auth_key;
+		hash_algo = oi->at_data.hash_algo;
+	}
+
+	if (!auth_str)
+		return -2;
+
+	hash_len = keychain_get_hash_len(hash_algo);
+	memset(apad, 0, sizeof(hash_len));
+	memset(temp_hash, 0, sizeof(hash_len));
+
+	/* start digest verification */
+	memcpy(apad, src, ipv6_addr_size);
+	memcpy(apad + ipv6_addr_size, ospf6_hash_apad_max,
+	       (hash_len - ipv6_addr_size));
+
+	auth_len = ntohs(ospf6_auth->length);
+
+	memcpy(temp_hash, ospf6_auth->data, hash_len);
+	memcpy(ospf6_auth->data, apad, hash_len);
+
+	ospf6_auth_update_digest(oi, oh, ospf6_auth, auth_str, auth_len,
+				 (oh_len + auth_len), hash_algo);
+
+#ifdef CRYPTO_OPENSSL
+	return !(CRYPTO_memcmp(temp_hash, ospf6_auth->data, hash_len));
+#else
+	return !(memcmp(temp_hash, ospf6_auth->data, hash_len));
+#endif
+}
+
+void ospf6_auth_digest_send(struct in6_addr *src, struct ospf6_interface *oi,
+			    struct ospf6_header *oh, uint16_t auth_len,
+			    uint32_t pkt_len)
+{
+	struct ospf6_auth_hdr *ospf6_auth;
+	char *keychain_name = NULL;
+	struct keychain *keychain = NULL;
+	struct key *key = NULL;
+	char *auth_str = NULL;
+	uint16_t key_id = 0;
+	enum keychain_hash_algo hash_algo = KEYCHAIN_ALGO_NULL;
+	uint32_t hash_len = KEYCHAIN_MAX_HASH_SIZE;
+	unsigned char apad[hash_len];
+	int ipv6_addr_size = sizeof(struct in6_addr);
+
+	if (CHECK_FLAG(oi->at_data.flags, OSPF6_AUTH_TRAILER_KEYCHAIN)) {
+		if (CHECK_FLAG(oi->at_data.flags,
+			       OSPF6_AUTH_TRAILER_KEYCHAIN_VALID)) {
+			auth_str = oi->at_data.auth_key;
+			hash_algo = oi->at_data.hash_algo;
+			key_id = oi->at_data.key_id;
+		} else {
+			keychain_name = oi->at_data.keychain;
+			keychain = keychain_lookup(keychain_name);
+			if (keychain) {
+				key = key_lookup_for_send(keychain);
+				if (key && key->string
+				    && key->hash_algo != KEYCHAIN_ALGO_NULL) {
+					auth_str = key->string;
+					hash_algo = key->hash_algo;
+					key_id = key->index;
+				}
+			}
+		}
+	} else if (CHECK_FLAG(oi->at_data.flags,
+			      OSPF6_AUTH_TRAILER_MANUAL_KEY)) {
+		auth_str = oi->at_data.auth_key;
+		hash_algo = oi->at_data.hash_algo;
+		key_id = oi->at_data.key_id;
+	} else {
+		if (IS_OSPF6_DEBUG_AUTH_TX)
+			zlog_warn("SEND[%s]: Authentication not configured for %s",
+				  oi->interface->name,
+				  ospf6_message_type(oh->type));
+		return;
+	}
+
+	if (!auth_str) {
+		if (IS_OSPF6_DEBUG_AUTH_TX)
+			zlog_warn("SEND[%s]: Authentication key is not configured for %s",
+				  oi->interface->name,
+				  ospf6_message_type(oh->type));
+		return;
+	}
+
+	hash_len = keychain_get_hash_len(hash_algo);
+	oi->at_data.seqnum_l++;
+
+	if (oi->at_data.seqnum_l == 0xFFFFFFFF) {
+		oi->at_data.seqnum_h++;
+		oi->at_data.seqnum_l = 0;
+	}
+
+	/* Key must be reset. which is not handled as of now. */
```

Comment:
Keys are also not stored, so if `ospf6d` is restarted we begin back, packets will be refused because the seqnum is too low.

Reply:
> We can't implement RFC 7166 without persistent seqnum storage; otherwise any `ospf6d` restart just breaks us since our neighbors will drop all packets.

This is not a blocker. 
In OSPF6 interfaces will persist until all configurations are deleted. 
when we have configured ospf6 with authentication trailer it wont delete the interface. hence previous sequence number will be intact. I know of this section of RFC.